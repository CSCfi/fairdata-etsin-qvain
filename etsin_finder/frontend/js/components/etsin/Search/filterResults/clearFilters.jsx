import React from 'react'
import styled from 'styled-components'
import { useHistory } from 'react-router-dom/cjs/react-router-dom.min'
import Translate from '@/utils/Translate'

import { useStores } from '@/stores/stores'
import { InvertedButton } from '@/components/etsin/general/button'

const ClearFilters = () => {
  const {
    Matomo,
    Accessibility,
    Etsin: {
      Search: { isLoading },
    },
    Locale: { translate },
  } = useStores()
  const history = useHistory()

  const clear = () => {
    Matomo.recordEvent('CLEAR_FILTERS')
    Accessibility.announce(translate('search.filter.filtersCleared'))
    history.push('/datasets')
  }

  if (isLoading) return null

  return (
    <Translate
      component={CustomButton}
      content="search.filter.clearFilter"
      onClick={clear}
      color="primary"
      open
    />
  )
}

const CustomButton = styled(InvertedButton)`
  display: ${p => (p.open ? 'initial' : 'none')};
  width: 100%;
  transition: 0.2s ease;
  margin: 0;
  margin-bottom: 0.5em;
`

export default ClearFilters
