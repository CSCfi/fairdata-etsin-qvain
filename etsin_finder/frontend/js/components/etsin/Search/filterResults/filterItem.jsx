import PropTypes from 'prop-types'
import { observer } from 'mobx-react'
import { useNavigate } from 'react-router-dom'

import { useStores } from '@/stores/stores'
import { useQuery } from '@/components/etsin/general/useQuery'

const quoteValue = value => {
  // Wrap value with quotation marks if needed
  if (value.includes(',') && !value.startsWith('"')) {
    return `"${value}"`
  }
  return value
}

const FilterItem = ({ filter, item, tabIndex }) => {
  const {
    Locale: { getValueTranslation },
  } = useStores()

  const query = useQuery()
  const navigate = useNavigate()
  const values = query.getAll(filter)

  const itemValue = quoteValue(getValueTranslation(item.value))
  const isActive = values.find(v => v === itemValue)

  function setFilter() {
    if (isActive) query.delete(filter, itemValue)
    else query.append(filter, itemValue)

    navigate(`/datasets?${query.toString()}`)
  }

  return (
    <li>
      <button
        role="switch"
        type="button"
        tabIndex={tabIndex}
        className={isActive ? 'active' : undefined}
        aria-checked={isActive}
        onClick={setFilter}
      >
        {`${getValueTranslation(item.value)} (${item.count})`}
      </button>
    </li>
  )
}

FilterItem.propTypes = {
  filter: PropTypes.string.isRequired,
  item: PropTypes.shape({
    value: PropTypes.shape({
      fi: PropTypes.string,
      en: PropTypes.string,
      und: PropTypes.string,
    }),
    count: PropTypes.number,
  }).isRequired,
  tabIndex: PropTypes.string.isRequired,
}

export default observer(FilterItem)
