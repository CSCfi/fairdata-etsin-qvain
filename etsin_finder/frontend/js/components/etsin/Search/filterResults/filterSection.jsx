import React, { useEffect, useState } from 'react'
import { observer } from 'mobx-react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faAngleDoubleUp, faAngleDoubleDown } from '@fortawesome/free-solid-svg-icons'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { mix } from 'polished'
import Translate from '@/utils/Translate'

import { TransparentLink } from '@/components/general/button'
import { useStores } from '@/stores/stores'

import FilterItem from './filterItem'
import { useQuery } from '@/components/etsin/general/useQuery'

const FilterSection = ({ filterName, onlyCurrentLanguage }) => {
  const {
    Etsin: {
      // isLoading will trigger update even though it is not strictly used here
      Search: { getAggregation, getAggregationQueryName, isLoading },
    },
    Locale: { getValueTranslation, lang },
  } = useStores()

  const query = useQuery()

  // sort activated items first
  const itemsSort = (a, b) => {
    const hasA = query.has(filter, getValueTranslation(a.value)) ? 1 : 0
    const hasB = query.has(filter, getValueTranslation(b.value)) ? 1 : 0
    if (hasB - hasA === 0) {
      // both activated or neither activated, sort descending
      return parseInt(b.count, 10) - parseInt(a.count, 10)
    }
    return hasB - hasA // sort activated first
  }

  const titleName = `search.aggregations.${filterName}.title`
  const filter = getAggregationQueryName(filterName)
  const items = getAggregation(filterName).slice().sort(itemsSort) || []
  let shownItems = items
  if (onlyCurrentLanguage) {
    shownItems = shownItems.filter(i => i.value[lang] || i.value.und)
  }
  const displayShowMoreButton = shownItems?.length > 10

  const [isOpen, setIsOpen] = useState(query.has(filter))
  const [showMoreItems, setShowMoreItems] = useState(false)

  useEffect(() => {
    if (!isLoading) {
      const f = getAggregationQueryName(filterName)
      if (query.has(f)) {
        setIsOpen(true)
      } else {
        setIsOpen(false)
      }
    }
  }, [setIsOpen, query, filterName, isLoading, getAggregationQueryName])

  const toggleSection = () => {
    setIsOpen(!isOpen)
  }

  if (displayShowMoreButton) {
    shownItems = showMoreItems ? shownItems : shownItems.slice(0, 9)
  }

  return (
    <Section>
      <Translate
        data-testid={`search-filter-${filterName}`}
        component={FilterCategory}
        onClick={toggleSection}
        aria-expanded={isOpen}
        content={`search.aggregations.${filterName}.title`}
      />
      <FilterItems className={isOpen ? 'open' : ''} aria-hidden={!isOpen}>
        <Translate component="ul" attributes={{ 'aria-label': titleName }}>
          {shownItems.map(item => (
            <FilterItem
              key={getValueTranslation(item.value)}
              filter={filter}
              item={item}
              tabIndex={isOpen ? '0' : '-1'}
            />
          ))}
        </Translate>
        {displayShowMoreButton && (
          <div>
            <hr />
            <ShowHide onClick={() => setShowMoreItems(!showMoreItems)}>
              <FontAwesomeIcon icon={showMoreItems ? faAngleDoubleUp : faAngleDoubleDown} />
              <ShowHideBtn>
                <Translate content={showMoreItems ? 'search.filter.hide' : 'search.filter.show'} />
              </ShowHideBtn>
            </ShowHide>
          </div>
        )}
      </FilterItems>
    </Section>
  )
}

FilterSection.propTypes = {
  filterName: PropTypes.string.isRequired,
  onlyCurrentLanguage: PropTypes.bool,
}

FilterSection.defaultProps = {
  // Enable to ignore items that don't have a translation in current language or "und" language
  onlyCurrentLanguage: false,
}

export default observer(FilterSection)

const ShowHide = styled.span`
  cursor: pointer;
`

const ShowHideBtn = styled(TransparentLink)`
  display: inline;
  margin-left: 8px;
`

export const Section = styled.li`
  margin-bottom: 4px;
`

// TODO: Better filter styles and animation
export const FilterCategory = styled.button`
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: space-between;
  color: ${p => p.theme.color.dark};
  border-radius: 0;
  width: 100%;
  text-align: left;
  border: 2px solid ${p => p.theme.color.lightgray};
  border-bottom: none;
  padding: 1em 1.5em;
  background-color: ${p => p.theme.color.lightgray};
  font-weight: 700;
  transition: all 0.3s ease;
  margin: 0;
  svg {
    pointer-events: none;
    height: 0.7em;
    display: flex;
    align-items: center;
  }
  &:focus,
  &:hover {
    background-color: ${p => mix(0.9, p.theme.color.lightgray, 'black')};
    border-color: ${p => mix(0.9, p.theme.color.lightgray, 'black')};
  }
  &:focus + .filter-items,
  &:hover + .filter-items {
    border-color: ${p => mix(0.9, p.theme.color.lightgray, 'black')};
  }
`

export const FilterItems = styled.div`
  padding: 0em 1em;
  max-height: 0px;
  overflow: hidden;
  transition: all 0.3s ease;
  ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
    transition: all 0.3s ease;
    li {
      max-height: 0;
      overflow: hidden;
      transition: all 0.3s ease;
      button:hover {
        text-decoration: underline;
        cursor: pointer;
      }
      button.active {
        background: ${p => p.theme.color.primary};
        color: ${p => p.theme.color.white};
      }
    }
  }
  &.open {
    max-height: 100%;
    padding: 1em 1em;
    li {
      max-height: 140px;
    }
  }
  button {
    background: transparent;
    border: none;
    padding: 0.3em 0.8em;
    border-radius: 0.7em;
    margin: 0 0 5px 0;
    color: ${p => p.theme.color.dark};
    text-align: left;
    &:focus {
      color: ${p => p.theme.color.primary};
      text-decoration: underline;
    }
  }
`
