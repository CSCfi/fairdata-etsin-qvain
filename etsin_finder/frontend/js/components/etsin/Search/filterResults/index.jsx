import styled from 'styled-components'
import { observer } from 'mobx-react'
import Translate from '@/utils/Translate'

import ClearFilters from './clearFilters'

import FilterSection from './filterSection'
import { useStores } from '@/stores/stores'

const FilterContainer = styled.ul`
  padding: 1em;
  margin-bottom: 1em;
  border: 2px solid ${props => props.theme.color.lightgray};
  @media (min-width: ${props => props.theme.breakpoints.lg}) {
    padding: 0em;
    border: none;
  }
`

const FilterResults = () => {
  const {
    Etsin: {
      Search: { isLoading },
    },
  } = useStores()

  if (isLoading) return null

  return (
    <FilterContainer aria-labelledby="filterlabel">
      <ClearFilters />
      <span id="filterlabel" className="sr-only" aria-hidden>
        <Translate content="search.filter.filters" />
      </span>
      <FilterSection filterName="data_catalog" />
      <FilterSection filterName="access_type" />
      <FilterSection filterName="organization" onlyCurrentLanguage />
      <FilterSection filterName="creator" onlyCurrentLanguage />
      <FilterSection filterName="field_of_science" />
      <FilterSection filterName="keyword" />
      <FilterSection filterName="infrastructure" />
      <FilterSection filterName="project" onlyCurrentLanguage />
      <FilterSection filterName="file_type" />
    </FilterContainer>
  )
}

export default observer(FilterResults)
