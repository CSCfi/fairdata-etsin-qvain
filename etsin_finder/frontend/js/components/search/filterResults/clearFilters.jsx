{
  /**
   * This file is part of the Etsin service
   *
   * Copyright 2017-2018 Ministry of Education and Culture, Finland
   *
   *
   * @author    CSC - IT Center for Science Ltd., Espoo Finland <servicedesk@csc.fi>
   * @license   MIT
   */
}
import React from 'react'
import styled from 'styled-components'
import Translate from 'react-translate-component'
import translate from 'counterpart'

import { InvertedButton } from '../../general/button'
import { useStores } from '../../../stores/stores'

const ClearFilters = () => {
  const { ElasticQuery, Matomo, Accessibility } = useStores()

  const clear = () => {
    Matomo.recordEvent('CLEAR_FILTERS')
    ElasticQuery.clearFilters()
    ElasticQuery.queryES()
    Accessibility.announce(translate('search.filter.filtersCleared'))
  }

  return (
    <CustomButton onClick={clear} color="primary" open={ElasticQuery.filter.length > 0}>
      <Translate content="search.filter.clearFilter" />
    </CustomButton>
  )
}

const CustomButton = styled(InvertedButton)`
  display: ${p => (p.open ? 'initial' : 'none')};
  width: 100%;
  transition: 0.2s ease;
  margin: 0;
  margin-bottom: 0.5em;
`

export default ClearFilters
