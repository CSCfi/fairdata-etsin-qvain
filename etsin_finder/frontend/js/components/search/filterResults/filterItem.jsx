{
  /**
   * This file is part of the Etsin service
   *
   * Copyright 2017-2018 Ministry of Education and Culture, Finland
   *
   *
   * @author    CSC - IT Center for Science Ltd., Espoo Finland <servicedesk@csc.fi>
   * @license   MIT
   */
}

import React, { Component } from 'react'
import PropTypes from 'prop-types'

import { withStores } from '../../../stores/stores'

class FilterItem extends Component {
  state = {
    term: this.props.term,
    key: this.props.item.key,
    doc_count: this.props.item.doc_count,
  }

  updateFilter = this.updateFilter.bind(this)

  // eslint-disable-next-line camelcase
  UNSAFE_componentWillReceiveProps(newProps) {
    this.setState({
      doc_count: newProps.item.doc_count,
      term: newProps.term,
    })
  }

  updateFilter() {
    const { ElasticQuery, Matomo } = this.props.Stores

    if (this.isActive()) {
      // will toggle to not active
      Matomo.recordEvent(`UNFILTER / ${this.props.sectionTitleEn} / ${this.state.key}`)
    } else {
      Matomo.recordEvent(`FILTER / ${this.props.sectionTitleEn} / ${this.state.key}`)
    }

    ElasticQuery.updateFilter(this.state.term, this.state.key)
    ElasticQuery.queryES()
  }

  isActive() {
    const { ElasticQuery } = this.props.Stores
    if (
      ElasticQuery.filter.filter(
        item => item.term === this.props.term && item.key === this.props.item.key
      ).length > 0
    ) {
      return true
    }
    return false
  }

  render() {
    return (
      <li>
        <button
          role="switch"
          type="button"
          tabIndex={this.props.tabIndex}
          className={this.isActive() ? 'active' : undefined}
          aria-checked={this.isActive()}
          onClick={() => {
            this.updateFilter()
          }}
        >
          {`${this.state.key} (${this.state.doc_count})`}
        </button>
      </li>
    )
  }
}

FilterItem.propTypes = {
  Stores: PropTypes.object.isRequired,
  term: PropTypes.string.isRequired,
  item: PropTypes.shape({
    key: PropTypes.string.isRequired,
    doc_count: PropTypes.number.isRequired,
  }).isRequired,
  tabIndex: PropTypes.string.isRequired,
  sectionTitleEn: PropTypes.string.isRequired,
}

export default withStores(FilterItem)
