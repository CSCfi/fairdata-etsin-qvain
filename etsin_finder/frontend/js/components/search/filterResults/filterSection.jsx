{
  /**
   * This file is part of the Etsin service
   *
   * Copyright 2017-2018 Ministry of Education and Culture, Finland
   *
   *
   * @author    CSC - IT Center for Science Ltd., Espoo Finland <servicedesk@csc.fi>
   * @license   MIT
   */
}

import React, { Component } from 'react'
import { observer } from 'mobx-react'
import Translate from 'react-translate-component'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faAngleDoubleUp, faAngleDoubleDown, faAngleDown } from '@fortawesome/free-solid-svg-icons'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { mix } from 'polished'

import FilterItem from './filterItem'
import { TransparentLink } from '../../general/button'
import { withStores } from '../../../stores/stores'

export class FilterSection extends Component {
  constructor(props) {
    super(props)

    // TODO: should this be in an app-level storage?
    this.aggregations = {
      access_type: {
        title: { en: 'Access', fi: 'Saatavuus' },
        aggregation: { en: 'access_type_en', fi: 'access_type_fi' },
        term: {
          en: 'access_rights.access_type.pref_label.en.keyword',
          fi: 'access_rights.access_type.pref_label.fi.keyword',
        },
      },
      organization: {
        title: { en: 'Organization', fi: 'Organisaatio' },
        aggregation: { en: 'organization_name_en', fi: 'organization_name_fi' },
        term: { en: 'organization_name_en.keyword', fi: 'organization_name_fi.keyword' },
      },
      creator: {
        title: { en: 'Creator', fi: 'TekijÃ¤' },
        aggregation: { und: 'creator' },
        term: { und: 'creator_name.keyword' },
      },
      field_of_science: {
        title: { en: 'Field of Science', fi: 'Tieteenala' },
        aggregation: { en: 'field_of_science_en', fi: 'field_of_science_fi' },
        term: {
          en: 'field_of_science.pref_label.en.keyword',
          fi: 'field_of_science.pref_label.fi.keyword',
        },
      },
      keyword: {
        title: { en: 'Keyword', fi: 'Avainsana' },
        aggregation: { en: 'all_keywords_en', fi: 'all_keywords_fi' },
        term: { en: 'all_keywords_en', fi: 'all_keywords_fi' },
      },
      infrastructure: {
        title: { en: 'Research Infra', fi: 'Tutkimusinfra' },
        aggregation: { en: 'infrastructure_en', fi: 'infrastructure_fi' },
        term: {
          en: 'infrastructure.pref_label.en.keyword',
          fi: 'infrastructure.pref_label.fi.keyword',
        },
      },
      project: {
        title: { en: 'Project', fi: 'Projekti' },
        aggregation: { en: 'project_name_en', fi: 'project_name_fi' },
        term: { en: 'project_name_en.keyword', fi: 'project_name_fi.keyword' },
      },
      file_type: {
        title: { en: 'File Type', fi: 'Tiedostotyyppi' },
        aggregation: { en: 'file_type_en', fi: 'file_type_fi' },
        term: {
          en: 'file_type.pref_label.en.keyword',
          fi: 'file_type.pref_label.fi.keyword',
        },
      },
      data_catalog: {
        title: { en: 'Data Catalog', fi: 'Datakatalogi' },
        aggregation: { en: 'data_catalog_en', fi: 'data_catalog_fi' },
        term: { en: 'data_catalog.en', fi: 'data_catalog.fi' },
      },
    }

    this.state = {
      open: this.props.filterOpen,
      show: false,
      aggregateItems: undefined,
      displayShowButton: undefined,
    }

    this.cutoff = 10
  }

  componentDidMount() {
    this.checkActive()
  }

  componentWillUnmount() {
    const SearchFilters = this.props.Stores.SearchFilters
    SearchFilters.closeFilters()
  }

  toggleFilter = () => {
    this.setState(state => ({
      open: !state.open,
    }))
  }

  toggleShowHideItems = () => {
    this.setState(state => ({
      show: !state.show,
    }))
  }

  checkIfValid = () => {
    if (
      this.aggregations[this.props.aggregation] === undefined ||
      this.props.Stores.ElasticQuery.results.aggregations[this.aggregationName] === 'undefined' ||
      this.props.Stores.ElasticQuery.results.aggregations[this.aggregationName].buckets.length <= 0
    ) {
      return false
    }
    return true
  }

  // Opens the section if it contains an active filter
  checkActive = () => {
    if (this.checkIfValid()) {
      this.props.Stores.ElasticQuery.results.aggregations[this.aggregationName].buckets.map(a => {
        if (!this.state.open) {
          const active =
            this.props.Stores.ElasticQuery.filter.filter(
              item => item.term === this.termName && item.key === a.key
            ).length > 0
          if (active) {
            this.setState({
              open: true,
            })
            return true
          }
        }
        return false
      })
    }
  }

  // Code that needs to be run before render()
  prepareForRender() {
    const {
      Locale: { getValueTranslation },
    } = this.props.Stores
    if (this.aggregations[this.props.aggregation] !== undefined) {
      const { title, aggregation, term } = this.aggregations[this.props.aggregation]
      this.titleName = getValueTranslation(title)
      this.aggregationName = getValueTranslation(aggregation)
      this.termName = getValueTranslation(term)
      this.titleNameEn = title.en
    }

    this.state.aggregateItems =
      this.props.Stores.ElasticQuery.results.aggregations[this.aggregationName].buckets
    this.state.displayShowButton = this.state.aggregateItems.length > this.cutoff
    if (!this.state.show) {
      this.state.aggregateItems = this.state.aggregateItems.slice(0, this.cutoff)
    }
  }

  render() {
    this.prepareForRender()
    const { results } = this.props.Stores.ElasticQuery
    if (results.total === 0) {
      return null
    }

    // Don't render unknown or empty
    if (!this.checkIfValid()) {
      return ''
    }
    return (
      <Section>
        <FilterCategory onClick={this.toggleFilter} aria-expanded={this.state.open}>
          {this.titleName}
          <FontAwesomeIcon icon={faAngleDown} size="2x" />
        </FilterCategory>
        <FilterItems className={this.state.open ? 'open' : ''} aria-hidden={!this.state.open}>
          <ul aria-label={this.titleName}>
            {this.state.aggregateItems.map(item => (
              <FilterItem
                key={item.key}
                item={item}
                sectionTitleEn={this.titleNameEn}
                term={this.termName}
                tabIndex={this.state.open ? '0' : '-1'}
              />
            ))}
          </ul>
          {this.state.displayShowButton ? (
            <div>
              <hr />
              <ShowHide onClick={this.toggleShowHideItems}>
                <FontAwesomeIcon icon={this.state.show ? faAngleDoubleUp : faAngleDoubleDown} />
                <ShowHideBtn>
                  <Translate
                    content={this.state.show ? 'search.filter.hide' : 'search.filter.show'}
                  />
                </ShowHideBtn>
              </ShowHide>
            </div>
          ) : (
            ''
          )}
        </FilterItems>
      </Section>
    )
  }
}

export default withStores(observer(FilterSection))

FilterSection.propTypes = {
  aggregation: PropTypes.string.isRequired,
  Stores: PropTypes.object.isRequired,
  filterOpen: PropTypes.bool,
}

FilterSection.defaultProps = { filterOpen: false }

const ShowHide = styled.span`
  cursor: pointer;
`

const ShowHideBtn = styled(TransparentLink)`
  display: inline;
  margin-left: 8px;
`

export const Section = styled.li`
  margin-bottom: 4px;
`

// TODO: Better filter styles and animation
export const FilterCategory = styled.button`
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: space-between;
  color: ${p => p.theme.color.dark};
  border-radius: 0;
  width: 100%;
  text-align: left;
  border: 2px solid ${p => p.theme.color.lightgray};
  border-bottom: none;
  padding: 1em 1.5em;
  background-color: ${p => p.theme.color.lightgray};
  font-weight: 700;
  transition: all 0.3s ease;
  margin: 0;
  svg {
    pointer-events: none;
    height: 0.7em;
    display: flex;
    align-items: center;
  }
  &:focus,
  &:hover {
    background-color: ${p => mix(0.9, p.theme.color.lightgray, 'black')};
    border-color: ${p => mix(0.9, p.theme.color.lightgray, 'black')};
  }
  &:focus + .filter-items,
  &:hover + .filter-items {
    border-color: ${p => mix(0.9, p.theme.color.lightgray, 'black')};
  }
`

export const FilterItems = styled.div`
  padding: 0em 1em;
  max-height: 0px;
  overflow: hidden;
  transition: all 0.3s ease;
  ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
    transition: all 0.3s ease;
    li {
      max-height: 0;
      overflow: hidden;
      transition: all 0.3s ease;
      button:hover {
        text-decoration: underline;
        cursor: pointer;
      }
      button.active {
        background: ${p => p.theme.color.primary};
        color: ${p => p.theme.color.white};
      }
    }
  }
  &.open {
    max-height: 100%;
    padding: 1em 1em;
    li {
      max-height: 140px;
    }
  }
  button {
    background: transparent;
    border: none;
    padding: 0.3em 0.8em;
    border-radius: 0.7em;
    margin: 0 0 5px 0;
    color: ${p => p.theme.color.dark};
    text-align: left;
    &:focus {
      color: ${p => p.theme.color.primary};
      text-decoration: underline;
    }
  }
`
