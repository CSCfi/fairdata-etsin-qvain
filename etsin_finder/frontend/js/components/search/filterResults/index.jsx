{
  /**
   * This file is part of the Etsin service
   *
   * Copyright 2017-2018 Ministry of Education and Culture, Finland
   *
   *
   * @author    CSC - IT Center for Science Ltd., Espoo Finland <servicedesk@csc.fi>
   * @license   MIT
   */
}

import React from 'react'
import styled from 'styled-components'
import Translate from 'react-translate-component'

import { observer } from 'mobx-react'
import FilterSection from './filterSection'
import { useStores } from '../../../utils/stores'

const FilterContainer = styled.ul`
  padding: 1em;
  margin-bottom: 1em;
  border: 2px solid ${props => props.theme.color.lightgray};
  @media (min-width: ${props => props.theme.breakpoints.lg}) {
    padding: 0em;
    border: none;
  }
`

const FilterResults = () => {
  const {
    SearchFilters: { fieldOfScienceIsOpen, keywordIsOpen, projectIsOpen },
  } = useStores()

  return (
    <FilterContainer aria-labelledby="filterlabel">
      <span id="filterlabel" className="sr-only" aria-hidden>
        <Translate content="search.filter.filters" />
      </span>
      <FilterSection aggregation="data_catalog" />
      <FilterSection aggregation="access_type" />
      <FilterSection aggregation="organization" />
      <FilterSection aggregation="creator" />
      <FilterSection aggregation="field_of_science" filterOpen={fieldOfScienceIsOpen} />
      <FilterSection aggregation="keyword" filterOpen={keywordIsOpen} />
      <FilterSection aggregation="infrastructure" />
      <FilterSection aggregation="project" filterOpen={projectIsOpen} />
      <FilterSection aggregation="file_type" />
    </FilterContainer>
  )
}

export default observer(FilterResults)
