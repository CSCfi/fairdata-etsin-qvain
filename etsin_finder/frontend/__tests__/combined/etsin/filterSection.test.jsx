import React from 'react'
import { shallow } from 'enzyme'
import {
  FilterSection,
  FilterCategory,
  Section,
  FilterItems,
} from '../../../js/components/search/filterResults/filterSection'
import { PasCheckBox } from '../../../js/components/search/filterResults/PasCheckBox'
import EnvClass from '../../../js/stores/domain/env'
import ElasticQueryClass from '../../../js/stores/view/elasticquery'

const Env = new EnvClass()
const mockElasticQuery = new ElasticQueryClass(Env)

const getStores = () => ({
  ElasticQuery: mockElasticQuery,
})

jest.mock('../../../js/stores/stores', () => {
  return {
    ...jest.requireActual('../../../js/stores/stores'),
    useStores: () => ({ ElasticQuery: mockElasticQuery }),
  }
})

const result1 = {
  hits: [
    {
      _index: 'metax',
      _type: 'dataset',
      _id: 'e8c85053-e5a4-4b13-a761-3d4263cb7962',
      _score: 12.766587,
      _source: {
        identifier: 'e8c85053-e5a4-4b13-a761-3d4263cb7962',
        data_catalog: {
          fi: 'Fairdata IDA-aineistot',
          sv: 'Fairdata forskningsdata',
          en: 'Fairdata IDA datasets',
        },
        data_catalog_identifier: 'urn:nbn:fi:att:data-catalog-ida',
        description: { fi: 'Kuvaus aineistoon  \nRivinvaihto  \n- bullet\n- bullet2' },
        access_rights: {
          license: [
            {
              identifier: 'http://uri.suomi.fi/codelist/fairdata/license/code/CC-BY-4.0',
              title: {
                fi: 'Creative Commons Nimeä 4.0 Kansainvälinen (CC BY 4.0)',
                en: 'Creative Commons Attribution 4.0 International (CC BY 4.0)',
                und: 'Creative Commons Nimeä 4.0 Kansainvälinen (CC BY 4.0)',
              },
            },
            {
              identifier: 'http://uri.suomi.fi/codelist/fairdata/license/code/CC-BY-2.0',
              title: {
                fi: 'Creative Commons Nimeä 2.0 Yleinen (CC BY 2.0)',
                en: 'Creative Commons Attribution 2.0 Generic (CC BY 2.0)',
                und: 'Creative Commons Nimeä 2.0 Yleinen (CC BY 2.0)',
              },
            },
          ],
          access_type: {
            identifier: 'http://uri.suomi.fi/codelist/fairdata/access_type/code/embargo',
            pref_label: { fi: 'Embargo', en: 'Embargo', und: 'Embargo' },
          },
        },
        title: { fi: 'Testi, Qvain webinar 20190912' },
      },
      highlight: { 'title.fi.finnish': ['<em>Testi</em>, Qvain webinar 20190912'] },
    },
  ],
  total: 1,
  aggregations: {
    project_name_en: { doc_count_error_upper_bound: 0, sum_other_doc_count: 0, buckets: [] },
    file_type_fi: {
      doc_count_error_upper_bound: 0,
      sum_other_doc_count: 0,
      buckets: [{ key: 'Kuva', doc_count: 1 }],
    },
    creator: {
      doc_count_error_upper_bound: 0,
      sum_other_doc_count: 0,
      buckets: [{ key: 'Erjan oma organisaatio', doc_count: 1 }],
    },
    data_catalog_fi: {
      doc_count_error_upper_bound: 0,
      sum_other_doc_count: 0,
      buckets: [{ key: 'Fairdata IDA-aineistot', doc_count: 1 }],
    },
    infrastructure_fi: { doc_count_error_upper_bound: 0, sum_other_doc_count: 0, buckets: [] },
    project_name_fi: { doc_count_error_upper_bound: 0, sum_other_doc_count: 0, buckets: [] },
    access_type_en: {
      doc_count_error_upper_bound: 0,
      sum_other_doc_count: 0,
      buckets: [{ key: 'Embargo', doc_count: 1 }],
    },
    file_type_en: {
      doc_count_error_upper_bound: 0,
      sum_other_doc_count: 0,
      buckets: [{ key: 'Image', doc_count: 1 }],
    },
    organization_name_en: {
      doc_count_error_upper_bound: 0,
      sum_other_doc_count: 0,
      buckets: [
        { key: 'CSC - IT Center for Science Ltd', doc_count: 1 },
        { key: 'Erjan oma organisaatio', doc_count: 1 },
      ],
    },
    all_keywords_en: {
      doc_count_error_upper_bound: 0,
      sum_other_doc_count: 0,
      buckets: [
        { key: 'cat', doc_count: 1 },
        { key: 'koira', doc_count: 1 },
        { key: 'lehmä', doc_count: 1 },
      ],
    },
    all_keywords_fi: {
      doc_count_error_upper_bound: 0,
      sum_other_doc_count: 0,
      buckets: [
        { key: 'kissa', doc_count: 1 },
        { key: 'koira', doc_count: 1 },
        { key: 'lehmä', doc_count: 1 },
      ],
    },
    organization_name_fi: {
      doc_count_error_upper_bound: 0,
      sum_other_doc_count: 0,
      buckets: [
        { key: 'CSC – Tieteen tietotekniikan keskus Oy', doc_count: 1 },
        { key: 'Erjan oma organisaatio', doc_count: 1 },
      ],
    },
    data_catalog_en: {
      doc_count_error_upper_bound: 0,
      sum_other_doc_count: 0,
      buckets: [{ key: 'Fairdata IDA datasets', doc_count: 1 }],
    },
    infrastructure_en: { doc_count_error_upper_bound: 0, sum_other_doc_count: 0, buckets: [] },
    access_type_fi: {
      doc_count_error_upper_bound: 0,
      sum_other_doc_count: 0,
      buckets: [{ key: 'Embargo', doc_count: 1 }],
    },
    field_of_science_en: {
      doc_count_error_upper_bound: 0,
      sum_other_doc_count: 0,
      buckets: [
        { key: 'Biomedicine', doc_count: 1 },
        { key: 'Environmental sciences', doc_count: 1 },
        { key: 'Nursing', doc_count: 1 },
      ],
    },
    field_of_science_fi: {
      doc_count_error_upper_bound: 0,
      sum_other_doc_count: 0,
      buckets: [
        { key: 'Biolääketieteet', doc_count: 1 },
        { key: 'Hoitotiede', doc_count: 1 },
        { key: 'Ympäristötiede', doc_count: 1 },
      ],
    },
  },
}

describe('FilterSection', () => {
  const stores = getStores()
  stores.ElasticQuery.results = result1
  const wrapper = shallow(<FilterSection aggregation="data_catalog" Stores={stores} />)
  it('should render <FilterSection />', () => {
    expect(wrapper).toMatchSnapshot()
  })
  it('should render <FilterCategory />', () => {
    expect(wrapper.find(Section).length).toBe(1)
  })
  it('should render <FilterCategory />', () => {
    expect(wrapper.find(FilterCategory).length).toBe(1)
  })
  it('should render <FilterItems />', () => {
    expect(wrapper.find(FilterItems).length).toBe(1)
  })
  it('should render <PasCheckBox />', () => {
    const wrapper = shallow(<PasCheckBox aggr="data_catalog" Stores={stores} />)
    expect(wrapper.find('#pasCheckbox').length).toBe(1)
  })
  it('should NOT render <PasCheckBox />', () => {
    const wrapper = shallow(<PasCheckBox aggr="asd" Stores={stores} />)
    expect(wrapper.find('#pasCheckbox').length).toBe(0)
  })
})
