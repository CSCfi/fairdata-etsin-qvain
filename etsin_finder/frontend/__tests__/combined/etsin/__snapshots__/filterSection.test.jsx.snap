// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterSection should render <FilterSection /> 1`] = `
<filterSection__Section>
  <filterSection__FilterCategory
    aria-expanded={false}
    onClick={[Function]}
  >
    Data Catalog
    <FontAwesomeIcon
      beat={false}
      beatFade={false}
      border={false}
      bounce={false}
      className=""
      fade={false}
      fixedWidth={false}
      flip={false}
      icon={
        Object {
          "icon": Array [
            448,
            512,
            Array [
              8964,
            ],
            "f107",
            "M201.4 374.6c12.5 12.5 32.8 12.5 45.3 0l160-160c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 306.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l160 160z",
          ],
          "iconName": "angle-down",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      maskId={null}
      pull={null}
      pulse={false}
      rotation={null}
      shake={false}
      size="2x"
      spin={false}
      spinPulse={false}
      spinReverse={false}
      swapOpacity={false}
      symbol={false}
      title=""
      titleId={null}
      transform={null}
    />
  </filterSection__FilterCategory>
  <filterSection__FilterItems
    aria-hidden={true}
    className=""
  >
    <ul
      aria-label="Data Catalog"
    >
      <ForwardRef
        item={
          Object {
            "doc_count": 1,
            "key": "Fairdata IDA datasets",
          }
        }
        key="Fairdata IDA datasets"
        sectionTitleEn="Data Catalog"
        tabIndex="-1"
        term="data_catalog.en"
      />
    </ul>
  </filterSection__FilterItems>
</filterSection__Section>
`;
